feat(context): add SetCookieData

Narita-1095305 opened the pull request:
# Add `SetCookieStruct` method to support flexible cookie setting

## Description
This PR adds a new `SetCookieStruct` method to the `Context` type that accepts a `*http.Cookie` pointer instead of individual parameters. This gives users more flexibility in setting cookie attributes, especially those not directly exposed by the current `SetCookie` method (such as `Expires`, `Partitioned`, etc).

The existing `SetCookie` method is preserved for backward compatibility.

## Changes
- Added `SetCookieStruct` method to `context.go`
- Added corresponding test cases in `context_test.go`
- Updated documentation in `doc.md` to demonstrate usage

## Related Issue
Resolves #4215

## Checklist
- [x] PR has only two commits
- [x] Added tests for the new functionality
- [x] Updated documentation
- [x] All CI tests pass

Narita-1095305 commented:
> Please rebase the master branch.

Rebased onto the latest master and updated the test assertion as suggested. Let me know if you have any further feedback!

appleboy commented on context.go:
```diff
@@ -1027,6 +1027,17 @@ func (c *Context) SetCookie(name, value string, maxAge int, path, domain string,
+// SetCookieStruct adds a Set-Cookie header to the ResponseWriter's headers.
+// It accepts a pointer to http.Cookie structure for more flexibility in setting cookie attributes.
+// The provided cookie must have a valid Name. Invalid cookies may be silently dropped.
+func (c *Context) SetCookieStruct(cookie *http.Cookie) {
```
Maybe rename to `SetCookieData` or something else?

appleboy commented on docs/doc.md:
```diff
@@ -2319,6 +2319,47 @@ func main() {
+  router := gin.Default()
+
+  router.GET("/cookie", func(c *gin.Context) {
+      cookie, err := c.Cookie("gin_cookie")
```
Replace four spaces with two spaces.

Narita-1095305 commented:
✅ Addressed the recent feedback:

- Renamed `SetCookieStruct` → `SetCookieData` as suggested
- Fixed indentation (replaced 4 spaces with 2 spaces)

Let me know if anything else needs adjustment. Thanks again!

Narita-1095305 commented:
✅ Addressed the review comments:
- Modified the logic to only apply `c.sameSite` when `cookie.SameSite` is `SameSiteDefaultMode`
- Updated tests to account for Go 1.22+ behavior where `SameSite=Lax` is omitted from the `Set-Cookie` header
